src/main/java/net/neoforged/neoforge/items/ItemHandlerHelper.java

```diff
@@ -185,4 +191,84 @@ public static int calcRedstoneFromInventory(@Nullable IItemHandler inv) {
+     * @param filter        a filter to select which items to move; if the filter returns {@code false} for a stack, it won't be moved
+     * @param maxAmount     maximum total amount of items to move
+     * @param stackInTarget {@code true} to try to stack with existing items in the target inventory,
+     *                      {@code false} to insert in the first slot that fits
+     * @return the total amount of moved items
+     */
+    public static int moveItems(
+            @Nullable IItemHandler from,
+            @Nullable IItemHandler to,
+            Predicate<ItemStack> filter,
+            int maxAmount,
+            boolean stackInTarget) {
+        Objects.requireNonNull(filter);
+        if (from == null || to == null || maxAmount <= 0) {
+            return 0;
+        }
+
+        int totalMoved = 0;
+
+        int fromSlots = from.getSlots();
+        slotsLoop:
+        for (int i = 0; i < fromSlots; ++i) {
+            // Check filter
+            if (!filter.test(from.getStackInSlot(i))) {
+                continue;
+            }
+
+            // Repeated extraction because extractItem limits to max stack size
+            while (true) {
+                // Simulate extraction
+                var available = from.extractItem(i, Integer.MAX_VALUE, true);
+                if (available.isEmpty()) {
+                    continue slotsLoop;
+                }
+                int availableCount = available.getCount();
+
+                // Simulate insertion
+                var simulationLeftover = stackInTarget ? insertItemStacked(to, available, true) : insertItem(to, available, true);
+                int canFit = availableCount - simulationLeftover.getCount();
+                if (canFit <= 0) {
+                    continue slotsLoop;
+                }
+
+                // Perform extraction
+                var extracted = from.extractItem(i, canFit, false);
+                if (extracted.isEmpty()) {
+                    continue slotsLoop;
+                }
+                int extractedCount = extracted.getCount();
+
+                // Perform insertion
+                var leftover = stackInTarget ? insertItemStacked(to, extracted, false) : insertItem(to, extracted, false);
+                int movedThisTime = extractedCount - leftover.getCount();
+                totalMoved += movedThisTime;
+
+                if (!leftover.isEmpty()) {
+                    // Try to give overflow back
+                    leftover = from.insertItem(i, leftover, false);
+
+                    if (!leftover.isEmpty()) {
+                        LOGGER.warn(
+                                "Trying to move up to {} items from {} to {}, but destination rejected {} that could not be inserted back in the source.",
+                                maxAmount, from, to, leftover);
+                    }
+                }
+
+                if (movedThisTime <= 0) {
+                    break;
+                }
```

Comment:
Since this is the only `break` in the loop, consider moving `movedThisTime` to outside the loop and turning this loop into a `do {} while ();` instead.

Reply:
I prefer not to, as it increases the scope of the variable unnecessarily, and doesn't particularly improve readability IMO. :D